### PR TITLE
Remove code originally deprecated in traits v3.0.3

### PR DESCRIPTION
**Pull request by rahulporuri**
This PR removes code which handles the `rich_compare` kwarg. We explicitly raise a `RuntimeWarning` now when we detect the `rich_compare` in the trait metadata during trait creation.

### CODE REVIEW

**Comment by mdickinson** [comment]
I'm not sure we're ready to remove this: I still see uses of `rich_compare` in blockcanvas, codetools, mayavi and scimath. Maybe we should introduce an explicit `DeprecationWarning` for uses of `rich_compare`, so that projects that use it get notified?

**Comment by mdickinson** [comment]
Could you please separate out the `repr` change into a different PR? I think that change is good to go.

**Comment by corranwebster** [comment]
Blockcanvas shouldn't prevent changes to Traits, as it isn't actively maintained.  Mayavi is probably the one that matters most out of the above.

**Comment by rahulporuri** [comment]
> Could you please separate out the repr change into a different PR? I think that change is good to go.

done in #599

**Comment by mdickinson** [comment]
#599 is merged to master; merging master into this branch should remove those changes from this PR.

**Comment by mdickinson** [comment]
All the existing uses of the `rich_compare` metadata in ETS have now been fixed (thanks, @rahulporuri)!

I'd like to go ahead and get rid of `rich_compare` in this PR; my main worry is that if there *is* code out there that's using `rich_compare`, this PR will result in a subtle and silent behaviour change, so will possibly be a cause of late-discovered bugs.

Suggested compromise: issue a `RuntimeWarning` in place of the existing code that handles `rich_compare`, if `rich_compare` is in the metadata. That'll at least ensure that users get a subtle and not-completely-silent behaviour change. We can remove the warning at some point later when everyone's had time to do the conversion.

**Comment by rahulporuri** [comment]
>  Suggested compromise: issue a RuntimeWarning in place of the existing code that handles rich_compare, if rich_compare is in the metadata. That'll at least ensure that users get a subtle and not-completely-silent behaviour change. We can remove the warning at some point later when everyone's had time to do the conversion.

@mdickinson - done

File: docs/source/traits_user_manual/defining.rst [hunk 1]
```diff
@@ -663,12 +663,18 @@ HasTraits objects:
...
+  generated based upon the result of comparing the old and new values of a
+  trait assignment:
+
+  * 0 (NO_COMPARE): The values are not compared and a trait change
+    notification is generated on each assignment.
+  * 1 (OBJECT_IDENTITY_COMPARE): A trait change notification is
+    generated if the old and new values are not the same object.
+  * 2 (RICH_COMPARE): A trait change notification is generated if the
+    old and new values are not equal using Python's
+    'rich comparison' operator. This is the default.
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst]: I'd just leave this as "are not equal." and lose the rest of the sentence. Python 3 only has rich comparisons, so there's only one obvious way to test for equality.

File: traits/traits.py [hunk 2]
```diff
@@ -1162,11 +1151,13 @@ def as_ctrait(self):
...
-        rich_compare = metadata.get("rich_compare")
-        if rich_compare is not None:
-            trait.rich_comparison(rich_compare is True)
+        # Ref : https://github.com/enthought/traits/issues/602
+        if "rich_compare" in metadata:
+            warnings.warn(
+                "The 'rich_compare' kwarg has been removed. Please "
+                "use the 'comparison_mode' kwarg instead.",
+                RuntimeWarning,
+            )
```
Comment (mdickinson) [inline traits/traits.py]: Would it be possible to add a test that exercises this branch?

File: traits/tests/test_rich_compare.py [hunk 3]
```diff
@@ -186,3 +189,20 @@ def test_assumptions(self):
...
+            class OldRichCompare(HasTraits):
+                bar = Any(rich_compare=False)
+
+        self.assertEqual(len(warn_msgs), 1)
+        warn_msg = warn_msgs[0]
+        self.assertIn(
+            "'rich_compare' kwarg has been removed",
+            str(warn_msg.message)
+        )
+        self.assertIn("trait_handlers", warn_msg.filename)
```
Comment (mdickinson) [inline traits/tests/test_rich_compare.py]: I don't think this is desirable: it's not useful for the user if the warning refers to the trait internals. Instead, the filename should point to the code that *uses* `rich_compare`, in this case this test file (`test_rich_compare.py`), so `"trait_handlers"` should be replaced with `"test_rich_compare"` and the `stacklevel` argument to the warning should be adjusted to whatever's necessary to make the test pass.

File: traits/trait_handlers.py [hunk 4]
```diff
@@ -659,11 +660,13 @@ def as_ctrait(self):
                 trait.post_setattr = post_setattr
                 trait.is_mapped(self.is_mapped)
 
-            # Note: The use of 'rich_compare' metadata is deprecated; use
-            # 'comparison_mode' metadata instead:
-            rich_compare = metadata.get("rich_compare")
-            if rich_compare is not None:
-                trait.rich_comparison(rich_compare is True)
+            # Ref : https://github.com/enthought/traits/issues/602
```
Comment (mdickinson) [inline traits/trait_handlers.py]: I think we can use the short form "enthought/traits#602" for issue references.

Comment (mdickinson) [reply traits/trait_handlers.py]: Gah. GitHub interpreted the markup. That should say: `enthought/traits#602`.

**Comment by mdickinson** [comment]
@rahulporuri Do you have bandwidth to update this PR (fix conflicts and address outstanding comments)?

**Comment by rahulporuri** [comment]
> @rahulporuri Do you have bandwidth to update this PR (fix conflicts and address outstanding comments)?

I do. Maybe not today but definitely by Monday.

**Comment by mdickinson** [comment]
I merged master in and resolved conflicts.

File: docs/source/traits_user_manual/defining.rst (hunk 1 above)
Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst]: This also now needs updating for the new enumerations.

Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst]: I've updated this to point to the `ComparisonMode` documentation, and avoided repeating the meaning of the values here.

File: traits/tests/test_rich_compare.py (hunk 3 above)
Comment (mdickinson) [reply traits/tests/test_rich_compare.py]: I've fixed this to check that the name of the module (so in this case "test_rich_compare") appears in the filename, and adjusted the stacklevel at the point where the `DeprecationWarning` is issue accordingly.

**Comment by mdickinson** [comment]
@rahulporuri I took another look at this. After some thought, I'm not comfortable with simply removing the `rich_compare` functionality, given that we know it *is* still used in the latest release of existing ETS packages like Mayavi.

I've reworked the PR to give explicit `DeprecationWarning`s rather than `RuntimeWarning`s, and to add additional tests. The `ctraits.c` function is still being removed, but uses of the `rich_compare` metadata are emulated by calling `comparison_mode` instead (and issuing the warning).

See what you think.

File: traits/traits.py (hunk 2 above)
Comment (mdickinson) [reply traits/traits.py]: Tests have been added in the rework.

File: traits/trait_handlers.py (hunk 4 above)
Comment (mdickinson) [reply traits/trait_handlers.py]: Done.

**Comment by rahulporuri** [comment]
> @rahulporuri I took another look at this. After some thought, I'm not comfortable with simply removing the rich_compare functionality, given that we know it is still used in the latest release of existing ETS packages like Mayavi.

I won't be able to work on this PR this week. Please feel free to update it/modify it as you see fit.

**Comment by mdickinson** [comment]
Merged master (post #715), and fixed up merge conflicts.

**Comment by mdickinson** [comment]
Merged master in again to resolve conflicts; the merge was somewhat messy.

File: traits/tests/test_traits.py [hunk 5]
```diff
@@ -1148,3 +1150,116 @@ def test_this_other_class(self):
...
+        )
+        _, _, this_module = __name__.rpartition(".")
+        self.assertIn(this_module, warn_msg.filename)
+
+    def test_rich_compare_false_warning(self):
+        with warnings.catch_warnings(record=True) as warn_msgs:
+            warnings.simplefilter("always", RuntimeWarning)
+
+            class A(HasTraits):
+                foo = Trait(rich_compare=True)
```
Comment (rahulporuri) [inline traits/tests/test_traits.py]: based on the test method name, this should be `rich_compare=False`

Comment (rahulporuri) [reply traits/tests/test_traits.py]: nvm. this was addressed in a newer commit. resolving.

Comment (mdickinson) [reply traits/tests/test_traits.py]: Hmm. I think you're commenting on code that's a couple of days old. :-(

**Comment by rahulporuri** [comment]
@mdickinson your changes LGTM. Thanks for the work and apologies for not addressing your comments myself.

**Pull request merged**